paasta_tools/docker_wrapper.py

```diff
@@ -71,31 +76,96 @@ def generate_hostname(fqdn, mesos_task_id):
     return hostname
 
 
-def add_hostname(args, hostname):
-    # Add --hostname argument immediately after 'run' command if it exists
+def add_argument(args, argument):
+    # Add an argument immediately after 'run' command if it exists
     args = list(args)
-
     try:
         run_index = args.index('run')
     except ValueError:
         pass
     else:
-        args.insert(run_index + 1, '--hostname=' + hostname)
-
+        args.insert(run_index + 1, argument)
     return args
 
 
+def get_cpumap():
+    # Return a dict containing the core numbers per physical CPU
+    core = 0
+    cpumap = {}
+    try:
+        with open('/proc/cpuinfo', 'r') as f:
+            for line in f:
+                m = re.match('physical\sid.*(\d)', line)
+                if m:
+                    cpuid = int(m.group(1))
+                    if cpuid not in cpumap:
+                        cpumap[cpuid] = []
+                    cpumap[cpuid].append(core)
+                    core += 1
+    except IOError:
+        pass
+    return cpumap
+
+
+def is_numa_enabled():
+    return os.path.exists('/proc/1/numa_maps')
+
+
+def arg_collision(new_args, current_args):
+    # Returns True if one of the new arguments is already in the
+    # current argument list.
+    cur_arg_keys = []
+    for c in current_args:
+        cur_arg_keys.append(c.split('=')[0])
+    return bool(set(new_args).intersection(set(cur_arg_keys)))
+
+
+def validate_int(i):
+    try:
+        return int(i)
+    except (ValueError, TypeError):
+        return None
+
+
+def validate_float(f):
+    # Ensure we return a float. If input is invalid we return 0.0
+    try:
+        return float(f)
+    except (ValueError, TypeError):
+        return 0.0
+
+
 def main(argv=None):
-    argv = argv if argv is not None else sys.argv[1:]
+    argv = argv if argv is not None else sys.argv
 
     env_args = parse_env_args(argv)
-    fqdn = socket.getfqdn()
 
     # Marathon sets MESOS_TASK_ID whereas Chronos sets mesos_task_id
     mesos_task_id = env_args.get('MESOS_TASK_ID') or env_args.get('mesos_task_id')
 
+    pinned_numa_node = validate_int(env_args.get('PIN_TO_NUMA_NODE'))
```

Comment:
This is all way too much to be in the "main" function.

Please extract as much as possible into a different function so that the code path in main is minimal. (only branch into numa stuff if the env var exists)